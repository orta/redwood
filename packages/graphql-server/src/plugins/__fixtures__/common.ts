import { makeExecutableSchema } from '@graphql-tools/schema'

export const testSchema = makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    type Query {
      me: User!
    }

    type Query {
      forbiddenUser: User!
      getUser(id: Int!): User!
    }

    type User {
      id: ID!
      name: String!
    }
  `,
  resolvers: {
    Query: {
      me: () => {
        return { _id: 1, firstName: 'Ba', lastName: 'Zinga' }
      },
      forbiddenUser: () => {
        throw Error('You are forbidden')
      },
      getUser: (id) => {
        return { id, firstName: 'Ba', lastName: 'Zinga' }
      },
    },
    User: {
      id: (u) => u._id,
      name: (u) => `${u.firstName} ${u.lastName}`,
    },
  },
})

export const testQuery = /* GraphQL */ `
  query meQuery {
    me {
      id
      name
    }
  }
`

export const testFilteredQuery = /* GraphQL */ `
  query FilteredQuery {
    me {
      id
      name
    }
  }
`

export const testErrorQuery = /* GraphQL */ `
  query forbiddenUserQuery {
    forbiddenUser {
      id
      name
    }
  }
`

export const testParseErrorQuery = /* GraphQL */ `
  query ParseErrorQuery {
    me {
      id
      name
      unknown_field
    }
  }
`

export const testValidationErrorQuery = /* GraphQL */ `
  query ValidationErrorQuery(id: Int!) {
    getUser(id: 'one') {
      id
      name
    }
  }
`
