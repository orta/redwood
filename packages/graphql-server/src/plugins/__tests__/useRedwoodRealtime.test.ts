import {
  createTestkit,
  createSpiedPlugin,
  assertStreamExecutionValue,
} from '@envelop/testing'

import { testLiveQuery, testSchema } from '../__fixtures__/common'
import {
  useRedwoodRealtime,
  InMemoryLiveQueryStore,
} from '../useRedwoodRealtime'

describe('useRedwoodRealtime', () => {
  const liveQueryStore = new InMemoryLiveQueryStore()

  it('should support a @live query directive', async () => {
    const testkit = createTestkit(
      [useRedwoodRealtime({ liveQueries: { liveQueryStore } })],
      testSchema
    )

    const result = await testkit.execute(testLiveQuery, {}, {})

    assertStreamExecutionValue(result)
    const current = await result.next()
    expect(current.value).toMatchInlineSnapshot(`
      {
        "data": {
          "me": {
            "id": "1",
            "name": "Ba Zinga",
          },
        },
        "isLive": true,
      }
    `)
  })

  it('should update schema with live directive', async () => {
    const spiedPlugin = createSpiedPlugin()

    // the original schema should not have the live directive before the useRedwoodRealtime plugin is applied
    expect(testSchema.getDirective('live')).toBeUndefined()

    createTestkit(
      [
        useRedwoodRealtime({ liveQueries: { liveQueryStore } }),
        spiedPlugin.plugin,
      ],
      testSchema
    )

    // the replaced schema should have the live directive afterwards
    const replacedSchema =
      spiedPlugin.spies.onSchemaChange.mock.calls[0][0].schema

    const liveDirectiveOnSchema = replacedSchema.getDirective('live')

    expect(liveDirectiveOnSchema.name).toEqual('live')
    expect(replacedSchema.getDirective('live')).toMatchSnapshot()
  })
})
