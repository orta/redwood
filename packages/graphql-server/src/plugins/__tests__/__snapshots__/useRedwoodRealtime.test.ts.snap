// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRedwoodRealtime should update schema with live directive 1`] = `"@live"`;
